pkg/rmb/rmb.go

```diff
@@ -0,0 +1,251 @@
+	TwinSrc    uint32   `json:"src"`
+	TwinDest   []uint32 `json:"dest"`
+	Retqueue   string   `json:"ret"`
+	Schema     string   `json:"shm"`
+	Epoch      int64    `json:"now"`
+	Err        string   `json:"err"`
+}
+
+// MessageBus is a struct that contains everything required to run the message bus
+type MessageBus struct {
+	Context  context.Context
+	pool     *redis.Pool
+	handlers map[string]func(ctx context.Context, payload []byte) (interface{}, error)
+}
+
+// New creates a new message bus
+func New(ctx context.Context, address string) (*MessageBus, error) {
+	pool, err := newRedisPool(address)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to connect to %s", address)
+	}
+
+	return &MessageBus{
+		pool:     pool,
+		Context:  ctx,
+		handlers: make(map[string]func(ctx context.Context, payload []byte) (interface{}, error)),
+	}, nil
+}
+
+// WithHandler adds a topic handler to the messagebus
+func (m *MessageBus) WithHandler(topic string, handler func(ctx context.Context, payload []byte) (interface{}, error)) {
+	m.handlers[topic] = handler
+}
+
+// Run runs listeners to the configured handlers
+// and will trigger the handlers in the case an event comes in
+func (m *MessageBus) Run() error {
+	con := m.pool.Get()
+	defer con.Close()
+
+	topics := make([]string, len(m.handlers))
+	for topic := range m.handlers {
+		topics = append(topics, topic)
+	}
+
+	jobs := make(chan Message, numWorkers)
+	for i := 1; i <= numWorkers; i++ {
+		go m.worker(context.Background(), jobs)
+	}
+
+	for {
+		if m.Context.Err() != nil {
+			return nil
+		}
+
+		data, err := redis.ByteSlices(con.Do("BLPOP", redis.Args{}.AddFlat(topics).Add(0)...))
+		if err != nil {
+			log.Err(err).Msg("failed to read from system local messagebus")
+			return err
+		}
+
+		var message Message
+		err = json.Unmarshal(data[1], &message)
+		if err != nil {
+			log.Err(err).Msg("failed to unmarshal message")
+			continue
+		}
+
+		_, ok := m.handlers[string(data[0])]
+		if !ok {
+			log.Debug().Msg("handler not found")
+			continue
+		}
+
+		jobs <- message
+	}
+}
+
+func (m *MessageBus) worker(ctx context.Context, jobs chan Message) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case message := <-jobs:
+			bytes, err := message.GetPayload()
+			if err != nil {
+				log.Err(err).Msg("err while parsing payload reply")
+			}
+
+			handler, ok := m.handlers[message.Command]
+			if !ok {
+				log.Warn().Msg("handler not found")
+			}
+
+			var tKey twinKeyID
+			ctx = context.WithValue(ctx, tKey, message.TwinSrc)
+
+			var mKey messageKey
+			ctx = context.WithValue(ctx, mKey, message)
+
+			data, err := handler(ctx, bytes)
+			if err != nil {
+				log.Err(err).Msg("err while handling job")
+				// TODO: create an error object
+				message.Err = err.Error()
+			}
+
+			err = m.SendReply(message, data)
+			if err != nil {
+				log.Err(err).Msg("err while sending reply")
+			}
+		}
+	}
+}
+
+func GetMessage(ctx context.Context) (*Message, error) {
```

Comment:
This should panic instead of returning an error. because failure to get the message is a programming error, not a runtime error. A failure to get the message meas the MessageBus never injected the message on the context. hence panic is much better here. 

It will also make the handler code much easier